torch/lib/TH/THStorage.c

```diff
@@ -66,39 +66,120 @@ TH_API THLongStorage *THLongStorage_newInferSize(THLongStorage *size, ptrdiff_t
+  ptrdiff_t ndim = dimsA > dimsB ? dimsA : dimsB;
 
   long *expandedSizes = THAlloc(sizeof(long)*ndim);
-  long *expandedStrides = THAlloc(sizeof(long)*ndim);
 
-  for (long i = numUnsqueezed; i < ndim; ++i) {
-    expandedSizes[i] = tensorSizes[i - numUnsqueezed];
-    expandedStrides[i] = tensorStrides[i - numUnsqueezed];
+  for (long i = ndim - 1; i >= 0; --i) {
+    long offset = ndim - 1 - i;
+    long dimA = dimsA - 1 - offset;
+    long dimB = dimsB - 1 - offset;
+    long sizeA = (dimA >= 0) ? sizesA[dimA] : 1;
+    long sizeB = (dimB >= 0) ? sizesB[dimB] : 1;
+    if (sizeA != sizeB) {
+      if (sizeA == 1) {
+        sizeA = sizeB;
+      }
+      else if (sizeB == 1) {
+      }
+      else {
+        THFree(expandedSizes);
+        if (raiseErrors) {
+          THError("The size of tensor a (%ld) must match the size of tensor b (%ld) at "
+                  "non-singleton dimension %ld.", sizeA, sizeB, i);
+        }
+        return -1;
+      }
+    }
+    expandedSizes[ i ] = sizeA;
   }
+  THLongStorage_resize(output, ndim);
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
+}
+
+TH_API int THLongStorage_inferSizeN(THLongStorage *output, int n, long **sizes, long *dims, int raiseErrors) {
+  THArgCheck(n > 0, 2, "n must be greater than 0");
+  THArgCheck(sizes != NULL, 1, "sizes must not be null");
+  THArgCheck(dims != NULL, 1, "dims must not be null");
 
-  for (long i = numUnsqueezed - 1; i > -1; --i) {
-    expandedSizes[i] = 1;
-    expandedStrides[i] = expandedSizes[i+1] * expandedStrides[i+1];
+  ptrdiff_t ndim = 0;
+  for (int j = 0; j < n; ++j) {
+    THArgCheck(sizes[ j ] != NULL, 1, "size %d must not be null", j);
+    THArgCheck(dims[ j ], 1, "Can't expand empty tensor %d", j);
+    ndim = dims[ j ] > ndim ? dims[ j ] : ndim;
   }
 
-  // create a new geometry for the tensor
-  for (long i = 0; i < ndim; ++i) {
-    long size = expandedSizes[i];
+  long *expandedSizes = THAlloc(sizeof(long)*ndim);
+
+  for (long i = ndim - 1; i >= 0; --i) {
+    long max_dim_size = 1;
+    long offset = ndim - 1 - i;
+    for (int j  = 0; j < n; ++j) {
+      long dim = dims[ j ] - 1 - offset;
+      long size = (dim >= 0) ? sizes[ j ][ dim ] : 1;
+      if (size != max_dim_size) {
+        if (max_dim_size == 1){
+          max_dim_size = size;
+        } else if (size == 1) {
+          // we'll expand, nothing to do
+        } else {
+          THFree(expandedSizes);
+          if (raiseErrors) {
+            THError("The size of tensor %i (%ld) must match the expanded size of tensor (%ld) at "
+                    "non-singleton dimension %ld.", j, size, max_dim_size, i);
+          }
+          return -1;
+        }
+      }
+    }
+
+    expandedSizes[ i ] = max_dim_size;
+  }
+  THLongStorage_resize(output, ndim);
+  memcpy(THLongStorage_data(output), expandedSizes, sizeof(long)*ndim);
+  THFree(expandedSizes);
+  return 0;
```

Comment:
Above in `THSize_isSameSizeAs` we return 1 on success, 0 on failure. Here we return 0 on success, -1 on failure? It would be nice for these to be consistent?